Source/x11/XGServerWindow.m

```diff
@@ -4409,75 +4415,152 @@ - (void) freecursor: (void*) cid
+   AppKit Screen - single monitor/display device.
+   X11 Screen - it's a plane where X Server can draw on.
+   XRandR - Xorg extension that helps to manipulate monitor layout providing
+   X11 Screen. 
+   We map XRandR monitors (outputs) to NSScreen. */
 - (NSArray *)screenList
 {
-  /* I guess screen numbers are in order starting from zero, but we
-     put the main screen first */
- int i;
- int count = ScreenCount(dpy);
- NSMutableArray *screens = [NSMutableArray arrayWithCapacity: count];
- if (count > 0)
-   [screens addObject: [NSNumber numberWithInt: defScreen]];
- for (i = 0; i < count; i++)
-   {
-     if (i != defScreen)
-       [screens addObject: [NSNumber numberWithInt: i]];
-   }
- return screens;
+  NSArray        *screens;
+  NSMutableArray *tmpScreens;
+  int            i;
+
+  monitorsCount = 0;
+  if (monitors != NULL) {
+    NSZoneFree([self zone], monitors);
+  }
+  
+#ifdef HAVE_XRANDR
+  XRRScreenResources  *screen_res;
+  RROutput            primary_output;
+  XRROutputInfo       *output_info;
+  XRRCrtcInfo         *crtc_info;
+
+  screen_res = XRRGetScreenResources(dpy, RootWindow(dpy, defScreen));
+  if (screen_res != NULL)
+    {
+      monitorsCount = screen_res->noutput;
+      tmpScreens = [NSMutableArray arrayWithCapacity: monitorsCount];
+      monitors = NSZoneMalloc([self zone], monitorsCount * sizeof(MonitorDevice));
+
+      primary_output = XRRGetOutputPrimary(dpy, [self xDisplayRootWindow]);
+      for (i = 0; i < monitorsCount; i++)
+        {
+          output_info = XRRGetOutputInfo(dpy, screen_res, screen_res->outputs[i]);
+          monitors[i].screen_id = defScreen;
+          if (output_info->crtc)
+            {
+              crtc_info = XRRGetCrtcInfo(dpy, screen_res, output_info->crtc);
+              
+                monitors[i].depth = [self windowDepthForScreen: i];
+                monitors[i].resolution = [self resolutionForScreen: defScreen];
+                monitors[i].frame = NSMakeRect(crtc_info->x, crtc_info->y,
+                                               crtc_info->width, crtc_info->height);
+                /* Add monitor ID (index in monitors array).
+                   Put primary monitor ID index 0 since NSScreen get this as main 
+                   screen if application has no key window. */
+                if (screen_res->outputs[i] == primary_output)
+                  {
+                    [tmpScreens insertObject: [NSNumber numberWithInt: i] atIndex: 0];
+                  }
+                else
+                  {
+                    [tmpScreens addObject: [NSNumber numberWithInt: i]];
+                  }
+            }
+        }
+      XRRFreeScreenResources(screen_res);
+      screens = [NSArray arrayWithArray: tmpScreens];
+    }
+#endif
+
+  if (monitorsCount == 0)
+    {
+      /* Assumed that it's always available 1 screen per application. We only
+         need to know it's number and it was saved in _initXContext as
```

Comment:
The second "it's" should be "its". This gets simpler if you turn the first one into "it is", or rather "there is" :-)

Reply:
Good point, thanks. Fixed.